### ISSUE TITLE
Optimize fact cache for scale

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#6807]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY
When the fact cache is utilized in AWX, we query the database for the fact blobs at job start for each host, write them to the execution bundle alongside the project, and send it off for execution.

This can lead to a notable startup penalty when a) the facts for a host are large b) the size of the inventory is large.

We should investigate what can be done to optimize this, both tactically at the in terms of optimizing queries, and strategically in terms of potentially moving to a different mode of fact storage.

**Comment by chrismeyersfsu** [comment]
The meat of the implementation is in [start_job_fact_cache](https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L830) and [finish_job_fact_cache](https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L855) To populate the Ansible fact cache we pull the per-host fact records out of the database and write them to individual `.json` files. AWX then enables the Ansible [jsonfile](https://docs.ansible.com/ansible/2.8/plugins/cache/jsonfile.html) fact cache plugin.

Updating/pushing facts back into the database is handled in `finish_job_fact_cache`. About the only "optimization" that we do is check to the see if the modification timestamp has changed on the per-host fact cache. If so, overwrite the old facts in the database for that host with the fact-cache version from the file.

#### Thoughts:
* Facts need to live in the database for the smart inventory feature.
* How slow is our current system? Why is it slow? Make an inventory with 10,000 hosts and default Ansible facts or 10x and 100x fact size.
  * Maybe there isn't an index on our search `hosts = hosts.filter(ansible_facts_modified__gte=timeout)`
  * Maybe getting all the fields for the host (i.e. the host vars) is slowing things down. Try `.values()`
  * Maybe making a file per-host and json dumping the data is slow?
  * Maybe Ansible loads the fact cache on init?
* edit: removed Redis idea. With the distributed nature of AWX and the direction we are going, we aren't going to able to lean on Redis.

**Comment by shanemcd** [comment]
Maybe helpful: https://docs.ansible.com/ansible/latest/dev_guide/developing_plugins.html#cache-plugins

**Comment by chrismeyersfsu** [comment]
https://github.com/ansible/awx/blob/devel/docs/debugging.md#code-instrumentation-timing Tooling that you can use to help you time various pieces of the fact cache process in AWX. Maybe you can stay in AWX without going into Ansible to know things about Ansible. For example, `playbook_on_start` is sort of the first event that is emitted from Ansible. Maybe you can rely on the `created` timestamp of that event to know when it happens in Ansible. You can observe the time between when the last thing happens in AWX and the `playbook_on_start` event `created` does not change as the size of the fact cache increases (either by increasing the number of hosts or increasing the per-host fact size). This time will be the Ansible start-up time. We can see if it changes or not.

Hopefully it doesn't come to this, but an example Ansible fact cache plugin that AWX used to use to interact with memcache https://github.com/ansible/awx/blob/1.0.4/awx/plugins/fact_caching/awx.py

Would be good to add the fact cache to the job lifecycle. Notifications are kind of in this same class of "work to do after an ansible job finishes" i.e. https://github.com/ansible/awx/blob/d7d496553e90dedfa0d318fd41398b3a3aaf2c7f/awx/main/tasks.py#L341

Once metrics lands. I wonder if it makes sense to add something from the fact cache as a metric https://github.com/ansible/awx/pull/9461/files

**Comment by dnguyen9** [comment]
@shanemcd @wenottingham Fact cache is used extensive for reporting purposes in our environment with  some inventories over 10K hosts.  A single job targetting 1 server with 'fact cache' enabled would generate about 2GB of facts from the database into the disk and memory. When we run multiple parallel jobs with fact cache enabled, we do eat up our memory and storage consumption pretty fast. 

In our environment, we just ended up modifying the behaviour of how facts are fetched in the  start_job_fact_cache and finish_job_fact_cache methods.  the rules are:
1. fetch facts for the servers that are specified in the --limit portion
2. if host limit is empty or does not return anything, return the entire facts table
3. _get_inventory_hosts() helper method did not appear to be used anywhere besides for start_job_fact_cache and finish_job_fact_cache to return the hosts objects associated with the inventory

this implementation seem to fix our issue. We applied the hotfix on 3/11 and saw behavior of memory consumption changed immediately. 
![image](https://user-images.githubusercontent.com/16950383/160636790-1184eb3d-edb0-4244-9d3f-cf266789ab91.png)


    def _get_inventory_hosts_hotfix(
        self,
        only=['name', 'ansible_facts', 'ansible_facts_modified', 'modified', 'inventory_id']
    ):
        '''
          HOTFIX_ID: 3870994
          Date : 9 Match 2022
          Reason: this customization allows us would like to pull only fact cache for hosts that are specified
          in the --limits options of AWX template with "enable fact_cache" so we don't have to pull about 
          2GB of data every time.


        '''

        if not self.inventory:
            return []
        hosts = self.inventory.hosts.only(*only)

        if self.limit:
            import re
            logger.debug("HOTFIX CUSTOM: Host limit passed by AWX: %s", self.limit)
            # basic regex filtering for host fqdn. does not support hostgroup or any advance targeting pattern
            list_limit = re.split(r'[,; \n]+', self.limit)
            hosts_new = hosts.filter(name__in=list_limit)
            # we match at least one host
            if hosts_new.count():
              logger.debug("HOTFIX CUSTOM: Filtered host list: %s", hosts_new)
              return hosts_new
            else:
              logger.debug('HOTFIX CUSTOM: Unable to find hosts with given limit provided. Returning facts for entire inventory')
        return hosts

**Comment by dnguyen9** [comment]
> The meat of the implementation is in [start_job_fact_cache](https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L830) and [finish_job_fact_cache](https://github.com/ansible/awx/blob/devel/awx/main/models/jobs.py#L855) To populate the Ansible fact cache we pull the per-host fact records out of the database and write them to individual `.json` files. AWX then enables the Ansible [jsonfile](https://docs.ansible.com/ansible/2.8/plugins/cache/jsonfile.html) fact cache plugin.
> 
> Updating/pushing facts back into the database is handled in `finish_job_fact_cache`. About the only "optimization" that we do is check to the see if the modification timestamp has changed on the per-host fact cache. If so, overwrite the old facts in the database for that host with the fact-cache version from the file.
> 
> #### Thoughts:
> * Facts need to live in the database for the smart inventory feature.
> * How slow is our current system? Why is it slow? Make an inventory with 10,000 hosts and default Ansible facts or 10x and 100x fact size.
>   
>   * Maybe there isn't an index on our search `hosts = hosts.filter(ansible_facts_modified__gte=timeout)`
>   * Maybe getting all the fields for the host (i.e. the host vars) is slowing things down. Try `.values()`
>   * Maybe making a file per-host and json dumping the data is slow?
>   * Maybe Ansible loads the fact cache on init?
> * edit: removed Redis idea. With the distributed nature of AWX and the direction we are going, we aren't going to able to lean on Redis.


From what I can tell, the database does a good job of storing our hosts table in memory and all subsequent calls are cached in memory so there wasn't a significant database penalty we saw. What we see as an issue are 1) the 2GB that we have to transfer between the DB and AWX node for each job that requires caching 2) the memory consumption since each job now holds a 2GB in memory as python dictionary 3) the json file is written to file so that also consumes storage. this gets annoying once we have 10K+ hosts. 

I would suggest that we configure the fact cache behaviour where it will only PULL facts for hosts that are in scope for the playbook. For example if limit is defined then return the facts for those hosts in the limit parameter.

**Comment by perzizzle** [comment]
I believe we are encountering this issue. We have ~6,500 hosts and noticed that if fact caching is enabled for a job template after launcing a job it would appear to spin for 1-2 minutes and then begin executing. If we disable fact caching, the jobs execute immediately.

**Comment by AlanCoding** [comment]
I put up a patch https://github.com/ansible/awx/pull/12998

> We should investigate what can be done to optimize this, both tactically at the in terms of optimizing queries, 

The patch does this.

> and strategically in terms of potentially moving to a different mode of fact storage.

The patch does not do this. 

> it would appear to spin for 1-2 minutes and then begin executing

It does not fix this, because the bottleneck there is probably disk interaction writing the facts. If it's the query, it's unlikely we can do much about it.

The patch does add a `.iterator()` so that hopefully, if it is taking a long time, it won't balloon memory to the extent mentioned in a prior comment here (assuming garbage collection works correctly).

**Comment by AlanCoding** [comment]
I'm going to close this in favor of https://github.com/ansible/awx/issues/12981

> https://github.com/ansible/awx/issues/12981

@ffirg you have made some comments that trended towards:

> and strategically in terms of potentially moving to a different mode of fact storage.

We could entertain ideas of that sort, but it's particularly challenging if this is meant to go through the receptor mesh and, as a consequence of that, get built into the ansible-runner transmit/worker/process contract. I'd just like to get another issue for that (if there's still a push for it) separate from these other very low-hanging fruit, quality-of-life improvements for fact caching.
